src/node/txorphanage.cpp

```diff
@@ -83,373 +216,484 @@ class TxOrphanageImpl final : public TxOrphanage {
-    // quantified in seconds and it is possible for orphans to have the same expiry.
-    std::sort(iters.begin(), iters.end(), [](const auto& lhs, const auto& rhs) {
-        if (lhs->second.nTimeExpire == rhs->second.nTimeExpire) {
-            return &(*lhs) < &(*rhs);
-        } else {
-            return lhs->second.nTimeExpire > rhs->second.nTimeExpire;
-        }
-    });
-    // Erase duplicates
-    iters.erase(std::unique(iters.begin(), iters.end()), iters.end());
-
-    // Convert to a vector of CTransactionRef
-    std::vector<CTransactionRef> children_found;
-    children_found.reserve(iters.size());
-    for (const auto& child_iter : iters) {
-        children_found.emplace_back(child_iter->second.tx);
-    }
     return children_found;
 }
 
 std::vector<TxOrphanage::OrphanTxBase> TxOrphanageImpl::GetOrphanTransactions() const
 {
-    std::vector<OrphanTxBase> ret;
-    ret.reserve(m_orphans.size());
-    for (auto const& o : m_orphans) {
-        ret.push_back({o.second.tx, o.second.announcers});
+    std::vector<TxOrphanage::OrphanTxBase> result;
+    result.reserve(m_unique_orphans);
+
+    auto& index_by_wtxid = m_orphans.get<ByWtxid>();
+    auto it = index_by_wtxid.begin();
+    std::set<NodeId> this_orphan_announcers;
+    while (it != index_by_wtxid.end()) {
+        this_orphan_announcers.insert(it->m_announcer);
+        // If this is the last entry, or the next entry has a different wtxid, build a OrphanTxBase.
+        if (std::next(it) == index_by_wtxid.end() || std::next(it)->m_tx->GetWitnessHash() != it->m_tx->GetWitnessHash()) {
+            result.emplace_back(it->m_tx, std::move(this_orphan_announcers));
+            this_orphan_announcers.clear();
+        }
+
+        ++it;
     }
-    return ret;
-}
+    Assume(m_unique_orphans == result.size());
 
-TxOrphanage::Usage TxOrphanageImpl::UsageByPeer(NodeId peer) const
-{
-    auto peer_it = m_peer_orphanage_info.find(peer);
-    return peer_it == m_peer_orphanage_info.end() ? 0 : peer_it->second.m_total_usage;
+    return result;
 }
 
 void TxOrphanageImpl::SanityCheck() const
 {
-    // Check that cached m_total_announcements is correct
-    TxOrphanage::Count counted_total_announcements{0};
-    // Check that m_total_orphan_usage is correct
-    TxOrphanage::Usage counted_total_usage{0};
-
-    // Check that cached PeerOrphanInfo::m_total_size is correct
-    std::map<NodeId, TxOrphanage::Usage> counted_size_per_peer;
-
-    for (const auto& [wtxid, orphan] : m_orphans) {
-        counted_total_announcements += orphan.announcers.size();
-        counted_total_usage += orphan.GetUsage();
-
-        Assume(!orphan.announcers.empty());
-        for (const auto& peer : orphan.announcers) {
-            auto& count_peer_entry = counted_size_per_peer.try_emplace(peer).first->second;
-            count_peer_entry += orphan.GetUsage();
+    std::unordered_map<NodeId, PeerDoSInfo> reconstructed_peer_info;
```

Comment:
d5097d6eef85eab971352cf78fb94a30c6e6f127

looks like you forgot to assert the calculated set matches anything ala

assert(m_peer_orphanage_info == reconstructed_peer_info);